modules/sdk-core/src/account-lib/mpc/tss/ecdsa/rangeproof.ts

```diff
@@ -0,0 +1,288 @@
+  curve: BaseCurve,
+  pk: PublicKey,
+  ntilde: NTilde,
+  c1: bigint,
+  c2: bigint,
+  x: bigint,
+  y: bigint,
+  r: bigint,
+  X: bigint
+): RangeProofWithCheck {
+  const q = curve!.order!();
+  const q3 = q ** BigInt(3);
+  const qntilde = q * ntilde.ntilde;
+  const q3ntilde = q3 * ntilde.ntilde;
+  const alpha = randBetween(q3);
+  const rho = randBetween(qntilde);
+  const sigma = randBetween(qntilde);
+  const tau = randBetween(qntilde);
+  const rhoprm = randBetween(q3ntilde);
+  const beta = randomCoPrimeTo(pk.n);
+  const gamma = randBetween(q3ntilde);
+  const u = curve!.basePointMult(curve!.scalarReduce(alpha));
+  const z = (modPow(ntilde.h1, x, ntilde.ntilde) * modPow(ntilde.h2, rho, ntilde.ntilde)) % ntilde.ntilde;
+  const zprm = (modPow(ntilde.h1, alpha, ntilde.ntilde) * modPow(ntilde.h2, rhoprm, ntilde.ntilde)) % ntilde.ntilde;
+  const t = (modPow(ntilde.h1, y, ntilde.ntilde) * modPow(ntilde.h2, sigma, ntilde.ntilde)) % ntilde.ntilde;
+  const v =
+    (((modPow(c1, alpha, pk._n2) * modPow(pk.g, gamma, pk._n2)) % pk._n2) * modPow(beta, pk.n, pk._n2)) % pk._n2;
+  const w = (modPow(ntilde.h1, gamma, ntilde.ntilde) * modPow(ntilde.h2, tau, ntilde.ntilde)) % ntilde.ntilde;
+  const hash = createHash('sha256');
+  hash.update('\x0d\x00\x00\x00\x00\x00\x00\x00');
+  hash.update(bigIntToBufferBE(pk.n));
+  hash.update('$');
+  hash.update(bigIntToBufferBE(pk.g));
+  hash.update('$');
+  hash.update(bigIntToBufferBE(X));
+  hash.update('$');
+  hash.update(bigIntToBufferBE(c1));
+  hash.update('$');
+  hash.update(bigIntToBufferBE(c2));
+  hash.update('$');
+  hash.update(bigIntToBufferBE(u));
+  hash.update('$');
+  hash.update(bigIntToBufferBE(z));
+  hash.update('$');
+  hash.update(bigIntToBufferBE(zprm));
+  hash.update('$');
+  hash.update(bigIntToBufferBE(t));
+  hash.update('$');
+  hash.update(bigIntToBufferBE(v));
+  hash.update('$');
+  hash.update(bigIntToBufferBE(w));
+  hash.update('$');
+  const e = bigIntFromBufferBE(hash.digest()) % q;
+  const s = (modPow(r, e, pk.n) * beta) % pk.n;
+  const s1 = e * x + alpha;
+  const s2 = e * rho + rhoprm;
+  const t1 = e * y + gamma;
+  const t2 = e * sigma + tau;
+  return { z, zprm, t, v, w, s, s1, s2, t1, t2, u };
+}
+
+/**
+ * Verify a zero-knowledge range proof that a homomorphically manipulated value is "small".
+ * @param {BaseCurve} curve An elliptic curve to use for group operations.
+ * @param {PublicKey} pk The prover's public key.
+ * @param {NTilde} ntilde The verifier's NTilde values.
+ * @param {RangeProofWithCheck} proof The range proof.
+ * @param {bigint} c1 The original ciphertext.
+ * @param {bigint} c2 The manipulated ciphertext.
+ * @param {bigint} X The curve's base point raised to x.
+ * @returns {boolean} True if verification succeeds.
+ */
+export function verifyWithCheck(
+  curve: BaseCurve,
+  pk: PublicKey,
+  ntilde: NTilde,
+  proof: RangeProofWithCheck,
+  c1: bigint,
+  c2: bigint,
+  X: bigint
+): boolean {
+  const q = curve!.order!();
```

Comment:
`!` shouldn't appear in production TypeScript.

It's not needed on `curve` and for `order` we should create a sub-interface and make these functions' signatures require curve be `BaseCurveWithOrder` (or whatever it ends up being called.

Reply:
It was easier to just make `order` a mandatory field and add the missing method to the ed25519 curve implementation.